From 84c4e8aad9137d0fd2a880d81da5cfe129ad4c16 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fw@strlen.de>
Date: Fri, 27 Mar 2020 03:24:46 +0100
Subject: [PATCH] netfilter: nf_queue: add nf_queue_entry_free() helper

This is a preparation patch, no logical changes.
Move free_entry into core and rename it to something more sensible.

Will ease followup patches which will complicate the refcount handling.

Signed-off-by: Florian Westphal <fw@strlen.de>
Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
(cherry picked from commit dd3cc111f2e3220ddc9c4ab17f13dc97759b5163)

[Quentin: keep nf_queue_entry_release_refs() exported for the kabi in case
it is used by out-of-tree modules, and make nf_queue_entry_free() also
free the nf_queue_entry_extra shadow variable.  Renamed from upstream's
"netfilter: nf_queue: make nf_queue_entry_release_refs static" as it no
longer makes it static.]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 include/net/netfilter/nf_queue.h |  1 +
 net/netfilter/nf_queue.c         |  8 ++++++++
 net/netfilter/nfnetlink_queue.c  | 11 ++---------
 3 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/include/net/netfilter/nf_queue.h b/include/net/netfilter/nf_queue.h
index 94f9e3a5aefd..42df66cac10d 100644
--- a/include/net/netfilter/nf_queue.h
+++ b/include/net/netfilter/nf_queue.h
@@ -56,6 +56,7 @@ void nf_reinject(struct nf_queue_entry *entry, unsigned int verdict);
 void nf_queue_entry_get_refs(struct nf_queue_entry *entry);
 bool nf_queue_entry_get_refs_bool(struct nf_queue_entry *entry);
 void nf_queue_entry_release_refs(struct nf_queue_entry *entry);
+void nf_queue_entry_free(struct nf_queue_entry *entry);
 
 static inline void init_hashrandom(u32 *jhash_initval)
 {
diff --git a/net/netfilter/nf_queue.c b/net/netfilter/nf_queue.c
index 754076e3ecb4..f89e1e7efa89 100644
--- a/net/netfilter/nf_queue.c
+++ b/net/netfilter/nf_queue.c
@@ -80,6 +80,14 @@ void nf_queue_entry_release_refs(struct nf_queue_entry *entry)
 }
 EXPORT_SYMBOL_GPL(nf_queue_entry_release_refs);
 
+void nf_queue_entry_free(struct nf_queue_entry *entry)
+{
+	nf_queue_entry_release_refs(entry);
+	shadow_var_free(entry, NF_QUEUE_ENTRY_EXTRA);
+	kfree(entry);
+}
+EXPORT_SYMBOL_GPL(nf_queue_entry_free);
+
 static void __nf_queue_entry_init_physdevs(struct nf_queue_entry *entry,
 					   struct nf_queue_entry_extra *extra)
 {
diff --git a/net/netfilter/nfnetlink_queue.c b/net/netfilter/nfnetlink_queue.c
index 3b3b77e649bc..647566bdcfe4 100644
--- a/net/netfilter/nfnetlink_queue.c
+++ b/net/netfilter/nfnetlink_queue.c
@@ -761,13 +761,6 @@ static void nf_bridge_adjust_segmented_data(struct sk_buff *skb)
 #define nf_bridge_adjust_segmented_data(s) do {} while (0)
 #endif
 
-static void free_entry(struct nf_queue_entry *entry)
-{
-	nf_queue_entry_release_refs(entry);
-	shadow_var_free(entry, NF_QUEUE_ENTRY_EXTRA);
-	kfree(entry);
-}
-
 static int
 __nfqnl_enqueue_packet_gso(struct net *net, struct nfqnl_instance *queue,
 			   struct sk_buff *skb, struct nf_queue_entry *entry)
@@ -793,7 +786,7 @@ __nfqnl_enqueue_packet_gso(struct net *net, struct nfqnl_instance *queue,
 		entry_seg->skb = skb;
 		ret = __nfqnl_enqueue_packet(net, queue, entry_seg);
 		if (ret)
-			free_entry(entry_seg);
+			nf_queue_entry_free(entry_seg);
 	}
 	return ret;
 }
@@ -854,7 +847,7 @@ nfqnl_enqueue_packet(struct nf_queue_entry *entry, unsigned int queuenum)
 
 	if (queued) {
 		if (err) /* some segments are already queued */
-			free_entry(entry);
+			nf_queue_entry_free(entry);
 		kfree_skb(skb);
 		return 0;
 	}
